### ISSUE TITLE
Bug: consistent-return does not recognise void operator in ternaries

### ISSUE THREAD

**Issue by robertrossmann** [eslint/eslint#8179]
**Tell us about your environment**

* **ESLint Version:** 3.16.1
* **Node Version:** 7.6
* **npm Version:** 4.3

**What parser (default, Babel-ESLint, etc.) are you using?** default

**Please show your full configuration:**

```
module.exports = {
  rules: {
    'consistent-return': [1. { treatUndefinedAsUnspecified: true }]
  }
}
```

**What did you do? Please include the actual source code causing the issue.**

```js
'use strict'

module.exports = function test(input) {
  if (input) {
    // Both paths are `void`ed, thus never returning a value
    return input === true
      ? void 0
      : void 1
  }

  // eslint error on next line: expected a return value
  return void false
}
```

**What did you expect to happen?**

Since all return paths a preceded by a `void` operator, I expected to see no error from the `consistent-return` rule because the function consistently returns `undefined` in all paths.

**What actually happened? Please include the actual, raw output from ESLint.**

ESLint complains about expecting a return value, likely because it failed to recognise the ternary branches as both being `void`ed.

```
test.js
  12:3  warning  Expected a return value  consistent-return

✖ 1 problem (0 errors, 1 warning)
```

**Comment by platinumazure** [comment]
Seems like a bug to me, hopefully someone can reproduce.

**Comment by not-an-aardvark** [comment]
This looks like an enhancement proposal to me, since it would improve the `undefined`-detecting behavior. Currently, it just detects `void` expressions that are directly returned.

It's worth noting that it's not possible to be completely consistent here. 

```js
const foo = unknownCondition ? bar : void 0;
return foo;
```

I suppose we could have the rule check for ternary expressions, but given that we're not going to be able to catch every case anyway, I wonder if it would be better to just have people use

```js
return void (foo ? bar : baz)
```

instead.

**Comment by robertrossmann** [comment]
I agree that we cannot statically catch all cases with 100% correctness. In the example above, I would expect that ESLint would warn on such code because for all intents and purposes, something (`foo`) is being returned.

However, if all return paths are `void`ed, there is no ambiguity. Therefore, I propose that we update this rule to also analyse returns in ternaries. Whether this is a bug or a feature request I leave to your own discretion.

I would argue that this is a bug because the `void` operator seems to be understood in most other situations (direct returns, switch/case, regular if/else etc.) with the ternary being the only one left out.

**Comment by nzakas** [comment]
Unfortunately, it looks like there wasn't enough interest from the team or community to implement this change. While we wish we'd be able to accommodate everyone's requests, we do need to prioritize. We've found that issues failing to be implemented after 90 days tend to never be implemented, and as such, we [close those issues](https://eslint.org/docs/maintainer-guide/issues#when-to-close-an-issue). This doesn't mean the idea isn't interesting or useful, just that it's not something the team can commit to.
